### PR TITLE
parse DataMessages as byte[] to fix Unicode issues

### LINKED ISSUE #17: Cannot process non-latin OTR encrypted messages from Pidgin/Psi+
Since Xabber switched to otr4j/otr4j implementation, it is not able to receive non-latin messages that are OTR encrypted from a Pidgin or Psi+ client, as reported by a user on https://github.com/redsolution/xabber-android/issues/385.

logcat gives the following (full logcat can be found on redsolution/xabber-android#385) :

```
I/System.out(12125): 11:51:41 AM RCV  (1111072848): <message id='purpleeee36e65' type='chat' to='xabberuser@server.com' from='pidginuser@server.com/1c3e9c54-2c2e-4575-994b-c9bbe19b9d88'><active xmlns='http://jabber.org/protocol/chatstates'/><body>?OTR:AAMDOdT/5wuLBjAAAAAAAQAAAAEAAADA65s1wprxbfsXHCBilc/OuQaoAlsRjQJGyVJn/qA4pgLIxiiymv6pz8YyBEaeLOtbHsW9pzWhBGfXafHp+JDuDVutduCI0K7TGRlYUL/WIYI7qRVuLCRwrWAJwH8hkN8YbxUoQdbUjITrNt2gyp+yDkDdHVTZlrF83LmxZhEfXyQ7HDvUEK54KCzRZTjc7p7xuqQzbYt67ZCVSsB72WWsTOqOtJBDJub+8aa/YoErMktbmxmdHrCOPJvKcRn7TfD4AAAAAAAAAAUAAABCQ0CityM6aunUNBAZoaookgP9st+J6lZCx14dUO8h7JpPLTCYM0Bqk40ptRywv8Bp75J2zfGmfRt5zzkpgevP+J50C7tfSVUIhCnIVKn6IYNSCucLKEQAAAAA.</body></message>
W/System.err(12125): xabberuser@server.com/androidrUmSh1VT:pidginuser@server.com
W/System.err(12125): net.java.otr4j.OtrException: java.io.IOException: Unable to read the required amount of bytes from the stream. Expected were 16718 bytes but I could only read 2 bytes.
W/System.err(12125):    at net.java.otr4j.session.Session.handleDataMessage(Session.java:643)
W/System.err(12125):    at net.java.otr4j.session.Session.transformReceiving(Session.java:455)
W/System.err(12125):    at com.xabber.android.data.extension.otr.OTRManager.transformReceiving(OTRManager.java:456)
W/System.err(12125):    at com.xabber.android.data.message.RegularChat.onPacket(RegularChat.java:144)
W/System.err(12125):    at com.xabber.android.data.message.MessageManager.onPacket(MessageManager.java:449)
```

### LINKED ISSUE #46: Cannot process non-latin OTR encrypted messages from Pidgin/Psi+
Since Xabber switched to otr4j/otr4j implementation, it is not able to receive non-latin messages that are OTR encrypted from a Pidgin or Psi+ client, as reported by a user on https://github.com/redsolution/xabber-android/issues/385.

logcat gives the following (full logcat can be found on redsolution/xabber-android#385) :
```
I/System.out(12125): 11:51:41 AM RCV  (1111072848): <message id='purpleeee36e65' type='chat' to='xabberuser@server.com' from='pidginuser@server.com/1c3e9c54-2c2e-4575-994b-c9bbe19b9d88'><active xmlns='http://jabber.org/protocol/chatstates'/><body>?OTR:AAMDOdT/5wuLBjAAAAAAAQAAAAEAAADA65s1wprxbfsXHCBilc/OuQaoAlsRjQJGyVJn/qA4pgLIxiiymv6pz8YyBEaeLOtbHsW9pzWhBGfXafHp+JDuDVutduCI0K7TGRlYUL/WIYI7qRVuLCRwrWAJwH8hkN8YbxUoQdbUjITrNt2gyp+yDkDdHVTZlrF83LmxZhEfXyQ7HDvUEK54KCzRZTjc7p7xuqQzbYt67ZCVSsB72WWsTOqOtJBDJub+8aa/YoErMktbmxmdHrCOPJvKcRn7TfD4AAAAAAAAAAUAAABCQ0CityM6aunUNBAZoaookgP9st+J6lZCx14dUO8h7JpPLTCYM0Bqk40ptRywv8Bp75J2zfGmfRt5zzkpgevP+J50C7tfSVUIhCnIVKn6IYNSCucLKEQAAAAA.</body></message>
W/System.err(12125): xabberuser@server.com/androidrUmSh1VT:pidginuser@server.com
W/System.err(12125): net.java.otr4j.OtrException: java.io.IOException: Unable to read the required amount of bytes from the stream. Expected were 16718 bytes but I could only read 2 bytes.
W/System.err(12125):    at net.java.otr4j.session.Session.handleDataMessage(Session.java:643)
W/System.err(12125):    at net.java.otr4j.session.Session.transformReceiving(Session.java:455)
W/System.err(12125):    at com.xabber.android.data.extension.otr.OTRManager.transformReceiving(OTRManager.java:456)
W/System.err(12125):    at com.xabber.android.data.message.RegularChat.onPacket(RegularChat.java:144)
W/System.err(12125):    at com.xabber.android.data.message.MessageManager.onPacket(MessageManager.java:449)
```

### PR DESCRIPTION
**Pull request by eighthave**
The original code was first parsing the message body into a `String`, then basing the TLV parsing off `String.length()`.  That works for ASCII, where one char is one byte, but totally fails for non-ASCII, since basically all non-latin chars are more than one byte.

This also has the advantage of removing the trailing null char from the end of the message body itself.

This should fix #46 for @aelmahmoudy

### CODE REVIEW

**Comment by aelmahmoudy** [comment]
Thanks ! Just tested Xabber with this patch for otr4j and it worked !

File: src/main/java/net/java/otr4j/session/Session.java [hunk 1]
```diff
@@ -619,13 +609,28 @@ private String handleDataMessage(DataMessage data) throws OtrException {
...
+                    if (dmc[i] == 0x00) {
+                        tlvIndex = i;
+                        break;
+                    }
+                }
+
+                // get the message body without the trailing 0x00, expect UTF-8 bytes
+                String decryptedMsgContent;
+                try {
+                    decryptedMsgContent = new String(dmc, 0, tlvIndex, "UTF-8");
```
Comment (cobratbq) [inline src/main/java/net/java/otr4j/session/Session.java]: If you use SerializationUtils.UTF8 here, you do not need to catch UnsupportedEncodingException. Makes the whole thing more readable.

**Comment by cobratbq** [comment]
Could you add a unit test for this. This is a great regression test for something that's hard to find otherwise.

Maybe we can extract the code for determining the start position of TVLs and move that to SerializationUtils?

File: src/main/java/net/java/otr4j/session/Session.java (hunk 1 above)
Comment (eighthave) [inline src/main/java/net/java/otr4j/session/Session.java]: Unfortunately, there is no equivalent `String` constructor that accepts a `CharSet`, which is what `SerializationUtils.UTF8` is.

Comment (eighthave) [inline src/main/java/net/java/otr4j/session/Session.java]: I changed it to use `SerializationUtils.UTF8.name()` which is as good as its going to get, I think.  But the `try` block will have to stay.

Comment (eighthave) [inline src/main/java/net/java/otr4j/session/Session.java]: Actually, I was totally wrong, switching to `SerializationUtils.UTF8`.  I looked for the right `String` constructor originally but for some reason, I concluded it wasn't there.

**Comment by eighthave** [comment]
As far as I can tell, this is the only spot in the code that needs to split off the TLVs from the rest of the message, so no need to move that stuff to a method.

As for unit tests, I have unicode related tests in another request: #48

**Pull request merged**